### PR TITLE
fix(suitability-triage): accept substantive AC checklists without a keyword

### LINKED ISSUE #2589: fix(suitability-triage): Check 7 requires a keyword even with a structured AC checklist
## Background

Refs #2586. `suitability-triage.mts`'s `checkVerifiability` (Check 7)
requires either `hasVerificationChannel`
(`/\btests?\b|\bverification\b|\bvalidate\b|\blint\b|\bci\b/i`) or
`hasObjectiveCriteria`. The latter, even when an
`## Acceptance Criteria` heading is followed by a real list or
checklist, additionally requires `OUTCOME_SIGNAL_PATTERN` to match
somewhere in that section or the whole body. `#2501` (closed) only
widened that keyword list's inflections — it did not change the
"a structured AC list alone is not enough" gate itself.

Reproduced on current `main` via `node scripts/suitability-triage.mjs
--stdin --verbose` against this body:

```
## Acceptance criteria
- [ ] `docs/example.md`'s pointer section reflects the confirmed owner
      and storage kind, with the open TODO callout removed.
- [ ] No confidential grant content appears anywhere in the updated
      page.
```

Check 5 (actionability) passes this body ("Issue defines actionable
scope and verifiable delivery details"). Check 7 (verifiability) fails
it ("Issue does not provide objective verification signals or
substantive acceptance criteria"), even though the checklist names a
specific file, a specific section, and two concrete, inspectable
outcomes. Adding one unrelated sentence containing a keyword from
either list anywhere in the body — with no change to the actual
acceptance criteria — flips Check 7 to pass.

This is distinct from `#2501` (inflections of the existing closed
list), `#2205` (subjective-approval proximity patches), and `#2512`
(the proximity gate on background prose). The remaining gap is that
Check 7's "substantive AC" path is a closed-vocabulary keyword
spot-check even when a genuine, structured checklist is already
present — disagreeing with Check 5, which already accepted the same
list as actionable.

## Proposed change

Once an `## Acceptance Criteria` heading (or an equivalent recognized
heading) is followed by a non-empty Markdown list or checklist, treat
that alone as satisfying `hasObjectiveCriteria`, without also
requiring `OUTCOME_SIGNAL_PATTERN` to match. Keep the existing keyword
lists (`hasVerificationChannel`, and `OUTCOME_SIGNAL_PATTERN` as a
fallback) for a body that has no recognized AC section at all, or
whose AC section has no list structure. If a further quality gate is
wanted for a near-empty or single-word bullet, key it off bullet
substance (a file path, a command, a named artifact) rather than a
closed English vocabulary.

## Acceptance criteria

- `node scripts/suitability-triage.mjs --stdin` against the
  reproduction body above reports the `verifiability` check as
  `pass`, with no other change to the body.
- `node scripts/suitability-triage.mjs --stdin` against a body with no
  `## Acceptance Criteria` heading, and no `tests`/`verification`/
  `validate`/`lint`/`ci` keyword anywhere, still reports
  `verifiability` as `fail` (the no-AC-section fallback keeps working).
- A body whose `## Acceptance Criteria` section contains only an empty
  or placeholder bullet (no path, command, or named artifact) still
  reports `verifiability` as `fail`, distinct from a genuinely
  structured checklist.
- New unit test(s) in `tests/suitability-triage.test.mts` cover: a
  structured AC checklist with no outcome-signal keyword (passes), an
  AC section with only a placeholder bullet (still fails), and the
  existing keyword-based fallback for a body with no AC section
  (unchanged).
- `pnpm run test -- suitability-triage` and `pnpm run lint:minimum`
  pass.

## Candidate files

- `src/scripts/suitability-triage.mts`
- `tests/suitability-triage.test.mts`

---

_Autopilot suitability: 4 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented._

---

_Effort: S -- author-estimated size; a soft autopilot selection
tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
# Summary

`checkVerifiability` (Check 7) in `suitability-triage.mts` required an
`OUTCOME_SIGNAL_PATTERN` keyword to appear inside a `## Acceptance
Criteria` section even when that section already held a genuine,
structured Markdown checklist — disagreeing with Check 5
(actionability), which already accepted the same list as actionable.

A bullet under the `## Acceptance Criteria` heading is now treated as
substantive on its own when `hasSubstantiveBullet()` finds an
inline-code span with both a real structural signal (a slash, dot,
underscore, colon, or hyphen) and actual alphanumeric content, and a
leading token that doesn't match `PLACEHOLDER_LEAD_PATTERN`, or a bare
dotted filename, with the existing outcome-signal keyword kept as a
fallback for a list with neither. The AC section is bounded at the next
Markdown heading — including one indented up to three spaces, per
CommonMark — so a trailing sibling section (this repo's own
`## Candidate files` convention) can't leak substance into a
placeholder AC bullet above it. Both the substantive-bullet check and
the outcome-signal fallback are scoped strictly to the section's own
list-item lines, so ordinary prose inside the section (including a
lazy-continuation line) can't supply a keyword or a code span the list
items themselves don't have. A fenced code block inside the section is
masked before any of this runs, so an illustrative bullet or
heading-like line quoted only as an example can neither supply fake
substance nor falsely truncate the section — and the code-span scan
itself no longer lets an unmatched backtick on one bullet pair across
lines with an unmatched backtick on another.

## Linked issue

Closes #2589

## Follow-up issues (if any)

- [x] none

## Background / rationale (if material)

This PR went through seven review rounds — three from Copilot, two from
an internal E2 critique subagent, one PR-body reconciliation (no code
change), and one from CodeRabbit — each finding a genuine false-positive
gap in the substantive-bullet heuristic (or a stale description of it),
fixed with a regression test per round:

1. A bare slash-path alternative also matched an ordinary conjunction
   (`and/or`, `N/A`, `6/30`) — dropped in favor of backtick spans +
   bare dotted filenames, and the AC section was bounded at the next
   heading so a trailing `## Candidate files` list couldn't leak
   substance into a placeholder bullet above it.
2. The backtick alternative treated *any* code span as substantive, so
   a placeholder wrapped in backticks (`` `TODO` ``, `` `N/A` ``)
   wrongly passed — fixed by requiring a structural signal inside the
   span plus a placeholder-token exclusion.
3. Internal whitespace counted as a structural signal, so a
   *multi-word* backticked placeholder (`` `TODO later` ``) still slipped
   past a single-token exact match — fixed by dropping whitespace as a
   signal and checking only the span's leading token (adding a colon
   to the signal set for npm-script names like
   `` `pnpm run lint:minimum` ``).
4. An E2 critique subagent found that round 3's lead-token split never
   split on a period, underscore, or slash, so a placeholder glued
   directly to one of those (`` `TODO.` ``, `` `TODO_` ``) still passed,
   and the "TODO/FIXME" case documented as an accepted limitation
   turned out to be a broader gap (any slash-glued suffix). Fixed by
   replacing the split-then-exact-match approach with a single pattern
   (`PLACEHOLDER_LEAD_PATTERN`) anchored at the start of the raw
   content: a placeholder word must be immediately followed by a
   non-alphanumeric character or the end of the string. Verified
   against a 34-case adversarial matrix (every prior round's cases plus
   the newly-found ones) before adding tests, with zero mismatches —
   this also closes the "TODO/FIXME" gap with no special-casing, while
   still correctly passing a real identifier that merely starts with
   the same letters (`` `NASA-report.txt` ``, `` `nonexistent-file.txt` ``).
5. Copilot flagged that this PR body's own "known accepted limitation"
   text (about the pre-round-4 `` `TODO/FIXME` `` gap) was stale once
   round 4 landed — no code change, just this body being reconciled
   with the code it describes.
6. CodeRabbit found three genuine gaps in the round-4 heuristic: a
   punctuation-only backtick span (e.g. `` `/` `` alone, no alphanumeric
   content) still counted as substantive; `NEXT_HEADING_PATTERN` didn't
   recognize a sibling heading indented up to three spaces (still a
   valid ATX heading per CommonMark), so it failed to bound the AC
   section there; and a non-list-item prose line inside the AC section
   (including a lazy-continuation line) could still supply a code span
   or outcome-signal keyword the list items themselves lacked. Fixed by
   adding `CODE_SPAN_ALNUM_PATTERN`, making `NEXT_HEADING_PATTERN`
   indentation-tolerant, and adding `extractListItemLines()` to scope
   both checks to the section's own list-item lines.
7. A second internal E2 critique subagent (round 6's fixes had no prior
   critique pass of their own) found three more gaps, each empirically
   confirmed to regress against the pre-PR build: (a) the code-span
   scan's `` [^`]+ `` also matched a newline, so two unrelated
   placeholder bullets each carrying one *unmatched* backtick got paired
   across the line join `extractListItemLines()` introduces, faking a
   substantive span out of two placeholders; (b) a fenced example
   bullet was read as a real list item, leaking substance into a
   placeholder bullet above it; (c) the mirror bug — a heading-like line
   inside a fence falsely bounded the section, hiding a genuine bullet
   after the fence. Fixed by excluding a newline from the code-span
   match and by masking fence content (not inline spans) over the whole
   body before any slicing, reusing `markdown-code.mts`'s existing
   position-preserving mask with a newly-exported fence-only range
   finder. The same pass also found and pinned (no code change) that a
   real identifier starting with a reserved placeholder word, with no
   dotted extension, glued directly to punctuation — e.g.
   `` `WIP-tracker/config` `` — is misclassified as a placeholder;
   documented as an accepted trade-off below rather than fixed, since
   widening the pattern to admit it would re-admit round 3-4's
   `TODO-later` / `TODO/FIXME` false-positive gap (structurally
   identical: placeholder word + one delimiter + word).

A separate, pre-existing whole-body fallback path (`hasNumSteps` /
`hasChecklist` in `checkVerifiability`, untouched by this diff) shares
the same "prose can supply the missing signal" bug class as finding 3
above, but no reviewer flagged it and it sits outside this issue's
scope — left as-is rather than fixed opportunistically.

**Known accepted limitation**: `` `WIP-tracker/config` `` (a real
identifier starting with a reserved placeholder word, no dotted
extension, glued directly to a slash) fails Check 7's substantive-bullet
check — see finding 7 above for why this is intentional rather than a
bug to fix.

## IDD impact

- [ ] Instruction files changed
- [ ] Template files changed
- [ ] Helper scripts changed
- [ ] Config schema changed
- [ ] Security / credential / merge behavior changed

## Verification

- [x] `pnpm lint`
- [x] `pnpm test` (`node --test tests/*.test.mts`; suitability-triage
      suite at 296/296, `cli-entry-smoke` TDZ guard at 21/21, full repo
      suite at 5021/5021)
- [x] `node scripts/audit-docs.mjs --check`
- [ ] `pnpm run docs:sync:check` (no docs/instructions touched)
- [x] `node scripts/idd-doctor.mjs` (if applicable)

## Safety

- [x] No secret-bearing examples
- [x] No private repo names / URLs
- [x] Merge policy impact reviewed
- [x] Context budget impact reviewed

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved Acceptance Criteria checks to recognize substantive file names, paths, commands, and identifiers—even when outcome-related keywords are absent.
  - Prevented placeholder examples, ordinary conjunctions, and content from later sections from being incorrectly treated as valid criteria.
  - Preserved existing keyword-based validation behavior while improving accuracy for structured checklists.

- **Tests**
  - Added coverage for valid structured criteria, placeholder rejection, dotted identifiers, commands, and section boundaries.

### CODE REVIEW

**Comment by kurone-kito** [comment]
advisory-wait: claude-050b7c33 6aa1da845ec2dabbce8f05fe49d79e6be404e8bf 2026-09-04T17:07:18Z

File: src/scripts/suitability-triage.mts [hunk 1]
```diff
@@ -662,6 +662,18 @@ function isEnumeratedParentheticalEntry(
...
+// identifier) or a bare dotted filename -- is substantive on its own,
+// independent of OUTCOME_SIGNAL_PATTERN's closed English vocabulary. Check 5
+// (actionability) already accepts the same checklist as actionable; this
+// keeps Check 7 from re-gating it behind a keyword spot-check the
+// checklist's own content already satisfies. Deliberately has no bare
+// slash-path alternative: `\bfoo\/bar\b` alone also matches an ordinary
+// conjunction like "and/or" or "read/write" with no real path underneath --
+// a real file path in this repo's AC bullets is either backticked or ends
+// in a dotted extension, both already covered below.
+const SUBSTANTIVE_BULLET_PATTERN = /`[^`]+`|\b[\w-]{2,}\.[a-zA-Z]{1,5}\b/;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts]: **Accepted** — good catch. The backtick alternative treated any inline-code span as substantive, so a placeholder like `TODO`/`N/A` wrapped in backticks wrongly passed. Fixed in 6f654fb9: an inline-code span now also needs a structural signal (a slash, dot, underscore, hyphen, or internal whitespace) and must not exactly match a short placeholder-token list (TODO, TBD, N/A, NA, XXX, FIXME, WIP, PENDING, PLACEHOLDER, NONE, ASAP). Added a regression test covering four backticked-placeholder variants.

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-050b7c33 6f654fb9e102b6b1a62447e44e4498e2ec7afb06 2026-09-04T17:13:51Z claim:17b2023f-d5c2-4d13-844d-17bec4d958fc attempt:1

File: tests/suitability-triage.test.mts [hunk 2]
```diff
@@ -2550,6 +2550,122 @@ test('verifiability still fails a bullet list with no outcome-signal word at all
...
+      body: 'Update the onboarding doc with the new steps for new contributors.',
+    },
+  } as Context);
+  assert.equal(result.pass, false);
+});
+
+test('verifiability accepts a bare dotted filename with no backticks or keyword (#2589 review)', () => {
+  // Exercises SUBSTANTIVE_BULLET_PATTERN's second alternative in isolation
+  // (every other new test hits the pattern only via a backtick-wrapped
+  // path): a plain, unquoted filename token is substantive on its own.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply tests/suitability-triage.test.mts]: **Accepted** — fixed. Updated both stale comment references (test lines ~2624 and ~2637 at the time of your review) to name the actual implementation symbols (`hasSubstantiveBullet()`, `BARE_DOTTED_FILENAME_PATTERN`) instead of the removed `SUBSTANTIVE_BULLET_PATTERN`, in e8d9038f.

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-050b7c33 e8d9038f60cbd84cdf2130fb91304ef1db4c816a 2026-09-04T17:20:40Z claim:17b2023f-d5c2-4d13-844d-17bec4d958fc attempt:1

**Comment by kurone-kito** [comment]
**Rejected** — chatgpt-codex-connector[bot] did not review HEAD e8d9038f60cbd84cdf2130fb91304ef1db4c816a (Codex usage limits for code reviews reached); this is not a completed review (source: #issuecomment-5543912353)

**Comment by kurone-kito** [comment]
**Rejected** — coderabbitai[bot] did not review HEAD e8d9038f60cbd84cdf2130fb91304ef1db4c816a (review limit reached / rate limited); this is not a completed review (source: #issuecomment-5543912874)

**Comment by kurone-kito** [comment]
_claude-050b7c33: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-050b7c33 322f323abddb0ad6fce1a0c4833286ca2b51c66e 2026-09-04T17:38:49Z claim:17b2023f-d5c2-4d13-844d-17bec4d958fc attempt:1

File: tests/suitability-triage.test.mts [hunk 3]
```diff
@@ -2550,6 +2550,248 @@ test('verifiability still fails a bullet list with no outcome-signal word at all
...
+test('verifiability rejects a placeholder glued directly to punctuation with no separator (#2589 round 4, E2 critique)', () => {
+  // An E2 critique subagent caught that round 3's split-on-whitespace/
+  // colon/hyphen lead-token check never split on a period, underscore, or
+  // slash, so a placeholder glued directly to one of those -- with no
+  // separating character at all -- still supplied its own structural
+  // signal while evading the exact-match stoplist: "- [ ] `TODO.`",
+  // "- [ ] `TODO_`", and the previously-documented "TODO/FIXME" gap (shown
+  // here to be broader than described: any slash-glued suffix, not just a
+  // second placeholder word). PLACEHOLDER_LEAD_PATTERN's anchored,
+  // non-alphanumeric-or-end lookahead closes all of these in one rule.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply tests/suitability-triage.test.mts:2684]: **Accepted** — the PR description was already out of date (fixed in a body edit shortly after 322f323a landed, per round 4). Round 4's anchored PLACEHOLDER_LEAD_PATTERN actually closes the TODO/FIXME gap entirely — verified both `TODO/FIXME` and `TODO/details forthcoming` now fail (see the regression test added in 322f323a). The description now says round 4 "closes the TODO/FIXME gap with no special-casing" instead of listing it as a limitation.

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-050b7c33 0432d38a80704b753c41041690f7262461b542d9 2026-09-04T17:42:32Z claim:17b2023f-d5c2-4d13-844d-17bec4d958fc attempt:1

File: src/scripts/suitability-triage.mts [hunk 4]
```diff
@@ -662,6 +662,60 @@ function isEnumeratedParentheticalEntry(
...
+// tried to reconstruct this via a whitespace/colon/hyphen split of the
+// span's lead token, which covered the punctuation marks it split on but
+// missed a placeholder glued directly to a period, underscore, or slash
+// -- an E2 critique subagent caught "TODO.", "TODO_", and the
+// previously-documented "TODO/FIXME" gap being broader than described.
+// Testing the anchored pattern against the raw content directly, rather
+// than a hand-split lead token, closes the whole class at once.
+const PLACEHOLDER_LEAD_PATTERN =
+  /^(?:TODO|TBD|N\/A|NA|XXX|FIXME|WIP|PENDING|PLACEHOLDER|NONE|ASAP)(?:[^a-zA-Z0-9]|$)/i;
+const CODE_SPAN_STRUCTURE_PATTERN = /[/._:-]/;
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts:712]: **Accepted** — fixed in 12b8a084. `hasSubstantiveBullet` now also requires the code span to contain at least one alphanumeric character (`CODE_SPAN_ALNUM_PATTERN`), not just the structural delimiter. Added a regression test covering `-`, `.`, `/`, `:`, `_` as the sole span content.

File: src/scripts/suitability-triage.mts [hunk 5]
```diff
@@ -1916,12 +1970,23 @@ export function checkVerifiability(context: Context): CheckOutcome {
...
-    if (/^[-*]\s+/.test(contentAfter) || /^\d+\.\s+/.test(contentAfter)) {
-      const hasOutcomeSignals = OUTCOME_SIGNAL_PATTERN.test(contentAfter);
-      if (hasOutcomeSignals) {
-        hasObjectiveCriteria = true;
-      }
+    // Bound the AC section at the next heading so a trailing sibling
+    // section (e.g. this repo's own "## Candidate files" convention, which
+    // is itself a bullet list of paths) never leaks substance or an
+    // outcome-signal keyword into a genuinely placeholder AC list (#2589).
+    const nextHeadingIndex = contentAfter.search(/\n#{1,6}\s/);
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts]: **Accepted** — fixed in 12b8a084. The heading-boundary pattern now allows up to three leading spaces before the `#` marker (`NEXT_HEADING_PATTERN`), matching CommonMark's ATX heading indentation allowance. Added a regression test with a 3-space-indented `## Candidate files` sibling section.

File: src/scripts/suitability-triage.mts [hunk 6]
```diff
@@ -1916,12 +1970,23 @@ export function checkVerifiability(context: Context): CheckOutcome {
...
+      nextHeadingIndex === -1
+        ? contentAfter
+        : contentAfter.slice(0, nextHeadingIndex);
+    // Require either a list (starting with - or *) or numbered content. A
+    // substantive bullet (hasSubstantiveBullet) satisfies this on its own;
+    // an outcome-signal keyword remains a fallback for a list that names
+    // no concrete file, command, or artifact (#2589).
+    if (/^[-*]\s+/.test(listSection) || /^\d+\.\s+/.test(listSection)) {
+      hasObjectiveCriteria =
+        hasSubstantiveBullet(listSection) ||
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts]: **Accepted** — fixed in 12b8a084. Both `hasSubstantiveBullet` and the outcome-signal keyword fallback now run only against `extractListItemLines()`'s output (lines that themselves start with a list marker), not the section's whole raw text. Added a regression test with a placeholder bullet followed by unrelated prose containing a real path.

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-050b7c33 12b8a084cd5e47fce38ede2899741d31bcef3212 2026-09-04T17:50:31Z claim:17b2023f-d5c2-4d13-844d-17bec4d958fc attempt:1

**Comment by kurone-kito** [comment]
_claude-050b7c33: review triage snapshot — IDD automation marker. Do not edit._

File: src/scripts/suitability-triage.mts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts:712]: **Accepted** — acknowledged; already fixed and resolved in round 6 (12b8a084).

File: src/scripts/suitability-triage.mts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts]: **Accepted** — acknowledged; already fixed and resolved in round 6 (12b8a084).

File: src/scripts/suitability-triage.mts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/suitability-triage.mts]: **Accepted** — acknowledged; already fixed and resolved in round 6 (12b8a084).

**Comment by kurone-kito** [comment]
_claude-050b7c33: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-050b7c33: critique baseline — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
advisory-wait-recovery: claude-050b7c33 d4791820639cf1e1d71a871a7b277d9581a01d1d 2026-09-04T18:20:27Z claim:17b2023f-d5c2-4d13-844d-17bec4d958fc attempt:1

**Comment by kurone-kito** [comment]
_claude-050b7c33: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-050b7c33: critique baseline — IDD automation marker. Do not edit._
